UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs

```diff
@@ -0,0 +1,209 @@
+
+            _commands = new GitUICommands(_moduleTestHelper.Module);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _moduleTestHelper.Dispose();
+        }
+
+        [Test]
+        public void PreserveCommitMessageOnReopen()
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+                formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+            });
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.AreEqual(generatedCommitMessage, formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [TestCase(CommitKind.Fixup)]
+        [TestCase(CommitKind.Squash)]
+        public void DoNotPreserveCommitMessageOnReopenFromSpecialCommit(CommitKind commitKind)
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(
+                formCommit =>
+                {
+                    string prefix = commitKind.ToString().ToLowerInvariant();
+                    Assert.AreEqual($"{prefix}! A commit message", formCommit.GetTestAccessor().Message.Text);
+                    formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+                },
+                commitKind);
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [Test]
+        public void SelectMessageFromHistory()
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(formCommit =>
+            {
+                var commitMessageToolStripMenuItem = formCommit.GetTestAccessor().CommitMessageToolStripMenuItem;
+
+                // Verify the message appears correctly
+                commitMessageToolStripMenuItem.ShowDropDown();
+                Assert.AreEqual("A commit message...", commitMessageToolStripMenuItem.DropDownItems[0].Text);
+
+                // Verify the message is selected correctly
+                commitMessageToolStripMenuItem.DropDownItems[0].PerformClick();
+                Assert.AreEqual("A commit message", formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        private static async Task WaitForIdleAsync()
+        {
+            var idleCompletionSource = new TaskCompletionSource<VoidResult>();
+            Application.Idle += HandleApplicationIdle;
+
+            // Queue an event to make sure we don't stall if the application was already idle
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            await Task.Yield();
```

Comment:
You know, we have this exact problem in VS where folks might schedule idle work, but idle is only executed on the transition from a non-empty queue to an empty queue. We call it "tickling idle" when you have to post a message just to get idle work to process again. 😄